### ISSUE TITLE
Gradient checkpointing for weight noise etc in PyTorch

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1552]
For variational noise (weight noise) or weight dropout or similar things, it would be very helpful to have gradient checkpointing where we can avoid that the weights are stored twice in memory, once with the noise, once without.

(Also see RF weight dropout discussion: #1518) 

See this current code as an example for variational noise in our TF code:
```python
if param_variational_noise and param.dtype.is_floating and isinstance(param, tf.Variable):
    with default_control_flow_ctx():  # make independent from loop/cond
        with reuse_name_scope_of_tensor(param, postfix="_variational_noise", add_tensor_name=True):

            def _apply_var_noise():
                rnd_state = tf_util.StatelessRandomSeed.create(shape=tf_util.get_shape(param))
                with gradient_checkpoint_scope():
                    noise = rnd_state.normal(stddev=param_variational_noise, dtype=param.dtype.base_dtype)
                    return param + noise

            param = self.network.cond_on_train(
                fn_train=_apply_var_noise,
                fn_eval=lambda: param,
            )
```
Specifically, check the code of `gradient_checkpoint_scope` and `prepare_gradient_checkpointing`.

There are also a couple of questions regarding how to implement dropout using gradient checkpointing and a stateless random number generator in RF and/or PT. E.g.:

* How to get stateless random numbers in PT? I think the best (only?) way is to create a new [`torch.Generator`](https://pytorch.org/docs/stable/generated/torch.Generator.html#torch.Generator)? I opened https://github.com/pytorch/pytorch/issues/128126 about this.
* Define a stateless random number API for RF? (Check maybe also JAX.)
* How to do the gradient checkpointing in RF? A new API for that? Currently `rf.dropout` is directly implemented using other RF methods, and this is nice and clean (you directly see the actual dropout implementation, it's not hidden away), and I think it should stay that way. But then this means we need to somehow extend this RF implementation by specifying the gradient checkpointing logic. And the gradient checkpointing should be optional.
  * See our own `gradient_checkpoint_scope` and `prepare_gradient_checkpointing` and how we implemented variational noise.
  * See [`torch.utils.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html) and [`tf.recompute_grad`](https://www.tensorflow.org/api_docs/python/tf/recompute_grad) for related functionality. Although I'm not sure that this API really allows for what we want (that the output is not stored in memory). But we could look at the internals of how it is done (specifically PyTorch), and maybe reuse that functionality to implement what we want.

**Comment by albertz** [comment]
I'm reading the code of `_checkpoint_without_reentrant_generator`. It looks like this uses a couple of techniques which are very relevant for what we need:

There is logic for `preserve_rng_state`. It gets the random state (CPU and device RNGs) and then `torch.random.fork_rng` ("Forks the RNG, so that when you return, the RNG is reset to the state that it was previously in.").

Further, it uses `torch.autograd.graph.saved_tensors_hooks` ([tutorial](https://pytorch.org/tutorials/intermediate/autograd_saved_tensors_hooks_tutorial.html)). That is the crucial functionality to control what tensors are stored for backprop, and how to reconstruct them.

The current [`torch.utils.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html) doesn't quite do what we want, as it does not control whether further ops *after* the `checkpoint` maybe store the tensor (which is what we want to avoid. I think one solution to our problem is to just defer the exit of the `saved_tensors_hooks` to some later point (when the tensor is not used further), because having `saved_tensors_hooks` active on the whole model probably gives some slowdown? I'm not exactly sure how to detect when it is not used anymore though. (Can I hook the tensor `__del__`?) And also, the whole logic around using `saved_tensors_hooks` in `checkpoint` seems quite involved and tricky (although the basic idea behind it is simple).

**Comment by albertz** [comment]
> Can I hook the tensor `__del__`?

One solution to this ([via](https://github.com/albertz/playground/blob/master/torch-tensor-track.py)):
```python
class _TensorHandle:
    def __init__(self, tensor):
        self.tensor_ref = weakref.ref(tensor)  # not really used here...

    def __del__(self):
        print("out-of-scope")


def func():
    x = torch.zeros(2, 3)
    x._my_handle = _TensorHandle(x)
    del x  # now (or later via GC) the _TensorHandle.__del__ will be called
    ...
```

**Comment by albertz** [comment]
I wonder a bit about [torch.utils.checkpoint](https://pytorch.org/docs/stable/checkpoint.html): Why do I need to specify the inputs for the function? What happens when I forget some? E.g. the example speaks about LSTM and passing `(activation, hidden)` to it. So another input which is missing here are the parameters. So what about them? They are not needed to be specified? Or what happens when you specify them? What happens when you don't specify them but still use them? If this does not matter, why specify the inputs at all?

**Comment by albertz** [comment]
I also wonder how [`torch.utils.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html) is supposed to be used in practice. The outputs of the checkpointing function are potentially anyway stored elsewhere for backprop, as those are not under the `checkpoint` anymore. E.g. consider the example:
```python
y = (a + b) * c
```
And now `checkpoint` around the `a + b`:
```python
a = ...
b = ...
c = ...
x = checkpoint(lambda (_a, _b): _a + _b, a, b)
y = x * c
```
It doesn't really make sense to recompute `x = a + b` here, because the result is anyway stored for the backprop of `x * c`. Or is the intended usage for this example actually like:
```python
a = ...
b = ...
c = ...
y = checkpoint(lambda (_a, _b, _c): (_a + _b) * c, a, b, c)
```
Or if explicit args are not necessary (see my previous comment):
```python
a = ...
b = ...
c = ...
y = checkpoint(lambda: (a + b) * c)
```

**Comment by albertz** [comment]
> I also asked the [question in the PyTorch discussion forum](https://discuss.pytorch.org/t/gradient-checkpointing/205416).

I got some quite useful answers which clarify my questions.

I have a very hacky idea: I can somehow delay the `_checkpoint_hook.__exit__` to some later point when `y._my_del_handler` or so goes out of scope...

**Comment by albertz** [comment]
Btw, I also have an idea on how to write a test case for this: We use some very similar code as in the example. Once without gradient checkpointing:
```python
x = a + b
y = x * c
```
Once with gradient checkpointing:
```
with gradient_checkpoint_scope():
    x = a + b
y = x * c
```
`a, b, c` should have `required_grad=True`.
And we record the memory consumption for both, using some of the Torch mechanisms (e.g. via [`torch.profiler`](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html) using `profile_memory=True`). Actually we only need to check memory consumption before the run and after the run (only counting Torch tensors), and it should be less for the second case. Actually exactly by the amount of tensor `x`. Or maybe we could also (maybe additionally) do a weakref on `x` and check if it is still alive in the second case.
Then we also should do backprop to verify that we get the same gradients in both cases.
The test could also be extended by using the random number generator, so see that the restoring of the RNG state works correctly.

**Comment by albertz** [comment]
Note, one problem with deferring the `saved_tensors_hook.__exit__`: If there are any other (unrelated) `saved_tensors_hook.__enter__` in the meantime, and then we call our `saved_tensors_hook.__exit__` while the new scope is active, this will lead to the wrong behavior.

Inside RETURNN, this is likely not so much a problem, as we don't really have any other `saved_tensors_hook` usage.

Note, one problem with putting any logic in `__del__`: This can potentially run in a different thread. Calling `saved_tensors_hook.__exit__` is thus wrong, as this logic is thread-local. It must execute in the same thread.

I'm not sure about the best solution. Or actually any solution which is not ugly.

I also reported that here: https://github.com/pytorch/pytorch/issues/129867

Currently the only solution I can think of is to hook into `saved_tensors_hook` itself (overwriting the `__enter__` and `__exit__` by own code), and add custom logic to handle all cases correctly.
